New elasticsearch command to update/rewrite data

dguendisch opened the pull request:
**What this PR does / why we need it**:
Adds a new command `elasticsearch` to update es data.
For now has only one subcommand `precompute`, which allows to rerun the computation of the precompute fields (stored in the teststep es documents in `.pre`) and update the data in elasticsearch.
Useful for modifying or introducing new pre fields and updating all existing data.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy|action
- target_group:   user|operator
-->
```improvement user
A new command `elasticsearch precompute` was introduced to allow to update all teststep `.pre` fields in elasticsearch to the current/most up-to-date precompute values.
```

schrodit commented on cmd/elasticsearch/cmd/precompute/precompute.go:
```diff
@@ -0,0 +1,171 @@
+	if err != nil {
+		return QueryResponse{}, err
+	}
+	logger.Log.V(6).Info("going to bulk update", "path", path)
```
shouldn't this be moved to line 163 to indicate the real update call?

dguendisch replied:
You're right!

schrodit commented on cmd/elasticsearch/cmd/precompute/precompute.go:
```diff
@@ -0,0 +1,171 @@
+
+
+	if !updateES {
+		logger.Log.Info("Not modifying elasticsearch data as update flag was not set")
```
maybe logging the build path and payload would also help debugging when doing a dry run

dguendisch replied:
Good idea, done.

schrodit commented on pkg/util/elasticsearch/elasticsearch.go:
```diff
@@ -68,11 +70,15 @@ func NewClient(cfg config.ElasticSearch) (Client, error) {
 
 func (c *client) Request(httpMethod, rawPath string, payload io.Reader) ([]byte, error) {
-	esURL, err := c.parseUrl(rawPath)
+	return c.RequestWithCtx(context.Background(), httpMethod, rawPath, payload)
```
this context should be canceled when the request is done
```suggestion
         ctx := context.Background()
         defer ctx.Done()
	return c.RequestWithCtx(ctx, httpMethod, rawPath, payload)
```

dguendisch replied:
Oh right, thanks for spotting, fixed.

schrodit commented on cmd/elasticsearch/cmd/precompute/precompute.go:
```diff
@@ -0,0 +1,171 @@
+		if oldPreComputed != nil {
+			clusterDomain = oldPreComputed.ClusterDomain
+		}
+		newPreComputed := collector.PreComputeTeststepFields(result.StepSummary.Phase, k8sVersion, clusterDomain)
```
you dont want to add a cluster domain if the old precomputed is not given.
Doesnt that mean that the cluster is only added if there are already precomputed fields?

dguendisch replied:
I struggled a bit to recompute that value in case this field is not already present (I'm not sure which data I can rely on for old datasets, so I don't want to set "some" clusterdomain there in case the information is not present).

schrodit commented on pkg/testmachinery/collector/precompute.go:
```diff
@@ -0,0 +1,68 @@
+	if clusterDomain != "" {
+		preComputed.ArgoDisplayName = "argo"
+		preComputed.LogsDisplayName = "logs"
+		preComputed.ClusterDomain = clusterDomain
```
the grafana and argo url are then computed like `argo.<clusterDomain>` ?

dguendisch replied:
Not really, the displayname is just a "display name" and makes an artificial column in the dataset, which can be used to attach computed values (like concatenations) in kibana/grafana at runtime. But one needs that static artificial column.

schrodit reviewed (changes requested):
just some minor observations

dguendisch commented on cmd/elasticsearch/cmd/precompute/precompute.go:
```diff
@@ -0,0 +1,171 @@
+	path, payload := BuildScrollQueryInitial("testmachinery-*", pageSize)
+
+	var processedItems int
+	for i := 0;; i++ {
```
👍

dguendisch commented:
Addressed your comments, thanks for them! Also added some better error handling/reporting in case bucket updates contain failures.
I now also added the first new precomputed field: `providerEnhanced`.
PTAL :)

schrodit reviewed (approved):
/lgtm
